source/user/watchdog-5.15-2/wd_keepalive.8
.TH WD_KEEPALIVE 8 "June 2015"
.UC 4
.SH NAME
wd_keepalive \- a simplified software watchdog daemon
.SH SYNOPSIS
.B wd_keepalive
.RB [ \-c " \fIfilename\fR|" \-\-config\-file " \fIfilename\fR]"
.SH DESCRIPTION
This is a simplified version of the
.B watchdog
daemon. Unlike the full watchdog, this daemon run no tests and only serves to keep
the hardware timer refreshed. Typically this is used on system start-up to
provide protection before the services that the full version tests are running,
and on shutdown to continue the refresh while those services are stopped.
.PP
If configured so it only opens
.IR /dev/watchdog ,
and keeps writing to it often enough to keep the kernel from resetting,
at least once per minute. Each write delays the reboot
time another minute. After a minute of inactivity the watchdog hardware will
cause a reset. In the case of the software watchdog the ability to
reboot will depend on the state of the machines and interrupts.
.PP
The wd_keepalive daemon can be stopped without causing a reboot if the device
.I /dev/watchdog
is closed correctly, unless your kernel is compiled with the
.I CONFIG_WATCHDOG_NOWAYOUT
option enabled.
.PP
Under high system load
.B wd_keepalive
might be swapped out of memory and may fail
to make it back in in time. Under these circumstances the Linux kernel will
reset the machine. To make sure you won't get unnecessary reboots make
sure you have the variable
.I realtime
set to
.I yes
in the configuration file
.IR watchdog.conf .
This adds real time support to
.BR wd_keepalive :
it will lock itself into memory and there should  be no problem even under the
highest of loads.
.PP
On system running out of memory the kernel will try to free enough memory by killing process. The
.B wd_keepalive
daemon itself is exempted from this so-called out-of-memory killer.
.SH OPTIONS
Available command line options are the following:
.TP
.BR \-c " \fIconfig-file\fR, " \-\-config\-file " \fIconfig-file"
Use
.I config-file
as the configuration file instead of the default
.IR /etc/watchdog.conf .
.TP
.BR \-X " \fInum\fR, " \-\-loop\-exit " \fInum"
Run for 'num' loops then exit as if SIG_TERM was received. Intended for test/debug (e.g. using
.B valgrind
for checking memory access). If the daemon exits on a loop counter and you have the
.I CONFIG_WATCHDOG_NOWAYOUT
option compiled for the kernel or device-driver then an unplanned reboot will follow - be warned!
.SH FILES
.TP
.I /dev/watchdog
The watchdog device.
.TP
.I /var/run/wd_keepalive.pid
The pid file of the running
.BR wd_keepalive .
.SH "SEE ALSO"
.BR watchdog.conf (5)
.TP
.BR watchdog (8)


source/History
Ralink AP SDK - History

Version 3.5.3.0
===============

- Support Ralink RT3883+RT5392 referenece board ( serial flash (SPI0) + vitesse switch (SPI1) )
- Fix mips 24kec "Wrong EPC on Exceptional J/JAL Targets" issue
- Fix raeth driver: hardware checksum offload only support ipv4 packets
- Fix igmpproxy bug: go on searching if it hits the CPU entry and consider WAN port is PPP0
- Fix static ip assigned issue of udhcpd
- Fix GUI bugs and function improvement
- Update ated: magic number check after wifi driver version 2.5

Version 3.5.2.0
===============

- Improve the WebUI
- Update the SPI CLK => BUS_CLK/8.
- Fix Ethernet driver compile error while enable RT3883 GMAC2
- Fix IPv6 neighbor solicitation issue for RT3352 and RT5350
- Fix the RT3352 Uboot memory size display error issue
- Fix the IGMP Proxy Makefile to support RT3352 and RT5350
- Fix RT5350 CPU/SYSCLK setting issue
- Fix RT3883 FLASH_CS0_CFG register for NOR flash controller
- RT5350 Uboot support CPU PLL adjustment
- RT3352/RT5350 support CPU Auto-Frequency adjustment
- RT3352/RT5350 Uboot support SDR pre-charge power down
- RT3352 Uboot support DDR self-refresh power save
- RT5350 ethernet LED control setting to optimize power save for Ralink reference board
- Add always check the erased word after erasing flash sector.
 
- Update RT3883 Uboot to turn ON external DRAM ODT as 150 ohm when write DRAM. (1000034C=0x452)
- Update RT3883 Uboot to turn ON RT3883 DDR IO ODT as 150 ohm when read DRAM. (10000014=0x03610180)

Version 3.5.0.0
===============

- support RT2880/RT3052/RT3883/RT3352 
- support NAND/SPI/NOR in the same firmware
- support Hardware NAT on RT3052/RT3883/RT3352
- support Software QoS (uplink/downlink)
- super dmz support
- support kernel mode pptp/l2tp to improve throughput significantly



source/user/watchdog-5.15-2/src/wd_keepalive.c
/**********************************************************
 * Copyright:   Appliance Studio Ltd
 *              Michael Meskes
 * License:	GPL
 *
 * Filename:    $Id: wd_keepalive.c,v 1.6 2007/08/17 09:24:54 meskes Exp $
 * Author:      Marcel Jansen, 22 February 2001
 * 		Michael Meskes, since then
 * Purpose:     This program can be run during critical periods
 *              when the normal watchdog shouldn't be run. It will
 *              read from the same configuration file, it will do
 *              no checks but will keep writing to the device
 *
***********************************************************/
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <limits.h>
#include <sched.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <linux/oom.h>
#include <linux/watchdog.h>
#include <libgen.h>
#include <string.h>
#include <unistd.h>

#include "extern.h"
#include "watch_err.h"

#define TRUE  1
#define FALSE 0

static void usage(char *progname)
{
	fprintf(stderr, "%s version %d.%d, usage:\n", progname, MAJOR_VERSION, MINOR_VERSION);
	fprintf(stderr, "%s [options]\n", progname);
	fprintf(stderr, "options:\n");
	fprintf(stderr, "  -c | --config-file <file>  specify location of config file\n");
	fprintf(stderr, "  -X | --loop-exit <number>  run a fixed number of loops then exit\n");
	exit(1);
}

/* Dummy function for keep_alive.c use */
int write_heartbeat(void)
{
	return 0;
}

/* close the device and check for error */
static void close_all(void)
{
	close_watchdog();
}

/* on exit we close the device and log that we stop */
void terminate(int ecode)
{
	log_message(LOG_NOTICE, "stopping watchdog keepalive daemon (%d.%d)", MAJOR_VERSION, MINOR_VERSION);
	unlock_our_memory();
	close_all();
	remove_pid_file();
	close_logging();
	usleep(100000);		/* 0.1s to make sure log is written */
	exit(ecode);
}

int main(int argc, char *const argv[])
{
	char *configfile = CONFIG_FILENAME;
	long count = 0L;
	long count_max = 0L;
	int c;
	char *progname;
	struct list *act;

	/* allow all options watchdog understands too */
	char *opts = "d:i:n:fsvbql:p:t:c:r:m:a:X:";
	struct option long_options[] = {
		{"config-file", required_argument, NULL, 'c'},
		{"force", no_argument, NULL, 'f'},
		{"sync", no_argument, NULL, 's'},
		{"no-action", no_argument, NULL, 'q'},
		{"verbose", no_argument, NULL, 'v'},
		{"softboot", no_argument, NULL, 'b'},
		{"loop-exit", required_argument, NULL, 'X'},
		{NULL, 0, NULL, 0}
	};

	progname = basename(argv[0]);
	open_logging(progname, MSG_TO_STDERR | MSG_TO_SYSLOG);

	/* check for the one option we understand */
	while ((c = getopt_long(argc, argv, opts, long_options, NULL)) != EOF) {
		switch (c) {
		case 'c':
			configfile = optarg;
			break;
		case 'n':
		case 'p':
		case 'a':
		case 'r':
		case 'd':
		case 't':
		case 'l':
		case 'm':
		case 'i':
		case 'f':
		case 's':
		case 'b':
		case 'q':
		case 'v':
			break;
		case 'X':
			count_max = atol(optarg);
			log_message(LOG_WARNING, "NOTE: Using --loop-exit so daemon will exit after %ld time intervals",
				    count_max);
			break;
		default:
			usage(progname);
		}
	}

	read_config(configfile);

	/* this daemon has no other function than writing to this device
	 * i.e. if there is no device given we better punt */
	if (devname == NULL) {
		log_message(LOG_INFO, " no watchdog device configured, aborting");
		exit(0);
	}

	if (wd_daemon(0, 0)) {
		fatal_error(EX_SYSERR, "failed to daemonize (%s)", strerror(errno));
	}

	/* tuck my process id away */
	if (write_pid_file(KA_PIDFILE)) {
		fatal_error(EX_USAGE, "unable to gain lock via PID file");
	}

	/* Log the starting message */
	open_logging(NULL, MSG_TO_SYSLOG);
	log_message(LOG_NOTICE, "starting watchdog keepalive daemon (%d.%d):", MAJOR_VERSION, MINOR_VERSION);
	log_message(LOG_INFO, " int=%d alive=%s realtime=%s", tint, devname, realtime ? "yes" : "no");

	/* open the device */
	if (open_watchdog(devname, dev_timeout) < 0) {
		terminate(1);
	}

	/* set signal term to call sigterm_handler() */
	/* to make sure watchdog device is closed */
	signal(SIGTERM, sigterm_handler);

	lock_our_memory(realtime, schedprio, daemon_pid);

	/* main loop: update after <tint> seconds */
	while (_running) {
		keep_alive();
		/* in pidmode kill -0 processes */
		for (act = pidfile_list; act != NULL; act = act->next) {
			if (check_pidfile(act) != ENOERR) {
				return 0;
			}
		}
		/* finally sleep some seconds */
		sleep(tint);

		count++;

		if (count_max > 0 && count >= count_max) {
			log_message(LOG_WARNING, "loop exit on interval counter reached");
			_running = 0;
		}
	}

	terminate(0);
	/* not reached */
	return 0;
}
